tools/iree-prof-output-utils.cc

```diff
@@ -8,75 +8,140 @@
 
 #include "third_party/abseil-cpp/absl/base/log_severity.h"
 #include "third_party/abseil-cpp/absl/flags/parse.h"
-#include "third_party/abseil-cpp/absl/log/check.h"
 #include "third_party/abseil-cpp/absl/log/globals.h"
 #include "third_party/abseil-cpp/absl/log/initialize.h"
+#include "third_party/abseil-cpp/absl/strings/str_cat.h"
 #include "third_party/abseil-cpp/absl/time/clock.h"
 #include "third_party/abseil-cpp/absl/time/time.h"
 #include "third_party/tracy/server/TracyWorker.hpp"
 
 namespace iree_prof {
+namespace {
 
-const char* GetZoneName(const tracy::Worker& worker,
-                        int16_t source_location_id) {
-  const auto& srcloc = worker.GetSourceLocation(source_location_id);
-  return worker.GetString(srcloc.name.active ? srcloc.name : srcloc.function);
+constexpr int kGpuThreadIndicator = 1 << 16;
+
+template <typename T>
+const T& GetEvent(const T& event) { return event; }
+
+template <typename T>
+const T& GetEvent(const tracy::short_ptr<T>& event) {
+  return *event;
 }
 
-void MergeDuration(int64_t start, int64_t end, std::vector<int64_t>& merged) {
-  CHECK_LE(start, end);
-  if (start == end) {
-    return;  // No changes.
+template <typename T>
+int64_t SumDurationInTimeline(const tracy::Worker& worker,
+                              const tracy::Vector<T>& timeline) {
+  int64_t duration = 0;
+  for (const auto& e : timeline) {
+    const auto& event = GetEvent(e);
+    duration += GetEventDuration(event);
   }
+  return duration;
+}
 
-  int least_index_greater_than_or_equal_to_start = 0;
-  while (least_index_greater_than_or_equal_to_start < merged.size() &&
-         start > merged[least_index_greater_than_or_equal_to_start]) {
-    ++least_index_greater_than_or_equal_to_start;
+uint64_t DecompressOrFixGpuThreadId(const tracy::Worker& worker,
+                                    uint16_t gpu_thread_id) {
+  if (gpu_thread_id < worker.GetThreadData().size()) {
```

Comment:
Please add a comment why this case and the else case happen.

Reply:
Done